specification/under_development/profiles/Adaptive_Profile.md

```diff
@@ -0,0 +1,985 @@
+  tail call void @__quantum__qis__x__body(%Qubit* nonnull inttoptr (i64 2 to %Qubit*))
+  br label %continue__1.i.i.i
+
+continue:
+...
+```
+
+Similarly to **Bullet 9**, the phi instruction can be used to move classical floating point values between branches.
+
+**Bullet 11: Classical fixed point computations computations** <br />
+If a backend has support for fixed point operations, they can also use the following intrinsics:
+Additionally if an adaptive profile program has support for floating point computations, the following instructions are supported:
+| LLVM Intrinsic | Context and Purpose               | Note                        |
+|:-----------------|:----------------------------------|:----------------------------|
+| `llvm.smul.fix.*`| Used to multiply two signed fixed point numbers.  |                             |
+| `llvm.smul.fix.sat*`| Same as above but clamps to min/max number in scale.  |                             |
+| `llvm.umul.fix.*`| Used to multiply two unsigned fixed point numbers. |                             |
+| `llvm.umul.fix.sat*`| Same as above but clamps to min/max number in scale.  |                             |
+| `llvm.sdiv.fix.*`| Used to divide two signed fixed point numbers. | Can cause real-time errors.                             |
+| `llvm.sdiv.fix.sat*`| Same as above but clamps to min/max number in scale.  |                             |
+| `llvm.udiv.fix.*`| Used to divide two unsigned fixed point numbers. | Can cause real-time errors. |
+| `llvm.udiv.fix.sat*`| Same as above but clamps to min/max number in scale.  |                             |
+|                  |                                   |                             |
+
+Similarly to **Bullet 9**, the phi instruction can be used to move classical fixed point values between branches.
+
+
+**Bullet 12: User defined functions and function calls** <br />
+An adaptive profile program may use user defined functions and function calls.
+For example, consider that with user defined functions if a back-end has a `Cnot` gate in its instruction set,
+then a program that liberally uses `Swap` operations can define a function and call it as follows:
+```llvm
+define void @swap(%Qubit* %arg1, %Qubit* %arg2) {
+call void __quantum__qis__cnot__body(%Qubit* %arg1, %Qubit* %arg2)
+call void __quantum__qis__cnot__body(%Qubit* %arg2, %Qubit* %arg1)
+call void __quantum__qis__cnot__body(%Qubit* %arg1, %Qubit* %arg2)
+}
+
+define void @main() {
+...
+call void @swap_0_1()
+...
+}
+```
+
+Moreover, classical functions can be defined in the IR assuming that a back-end has opted into supporting
+classical computations:
+
+```llvm
+define i64 @triple(i64 %0) {
+%1 = mul i64 %0, 3
+ret i64 %1
+}
+
+define void @main() {
+...
+%0 = call void @triple(i64 2)
+...
+}
+```
+
+The only restriction on user functions and definitions is that you cannot have dynamically allocated `%Qubit*`  arguments, they must still be constant `%Qubit*` id's. 
```

Comment:
I think the phrasing here requires some context due to the mention of dynamic qubit allocation. An alternative way of phrasing it with more details would be something like this:

There are three restrictions on user-defined functions if they operate on qubits:
1. If a user-defined function calls a quantum instruction set (QIS) function (e.g. a function with `__quantum__qis` prefix), the `Qubit%*` arguments to the QIS function must come from a `%Qubit*` arguments of the user-defined function itself.

**Valid example**:
```llvm
define void @swap(%Qubit* %arg1, %Qubit* %arg2) {
    call void __quantum__qis__cnot__body(%Qubit* %arg1, %Qubit* %arg2)
    call void __quantum__qis__cnot__body(%Qubit* %arg2, %Qubit* %arg1)
    call void __quantum__qis__cnot__body(%Qubit* %arg1, %Qubit* %arg2)
}
define void @main() {
...
    call void @swap(%Qubit* null, %Qubit* nonnull inttoptr (i64 1 to %Qubit*))
...
}
```

**Invalid example**:
```llvm
define void @swap() {
    ; This is invalid because __quantum_qis* functions are being called from within a user-defined function and the provided `%Qubit*` arguments are constant.
    call void __quantum__qis__cnot__body(%Qubit* null, %Qubit* nonnull inttoptr (i64 1 to %Qubit*))
    call void __quantum__qis__cnot__body(%Qubit* nonnull inttoptr (i64 1 to %Qubit*), %Qubit* null)
    call void __quantum__qis__cnot__body(%Qubit* null, %Qubit* nonnull inttoptr (i64 1 to %Qubit*))
}
define void @main() {
...
    call void @swap()
...
}
```

2. If called from the entry-point function, they must receive constant `%Qubit*` IDs as arguments.

**Valid example**:
```llvm
define void @swap(%Qubit* %arg1, %Qubit* %arg2) {
    call void __quantum__qis__cnot__body(%Qubit* %arg1, %Qubit* %arg2)
    call void __quantum__qis__cnot__body(%Qubit* %arg2, %Qubit* %arg1)
    call void __quantum__qis__cnot__body(%Qubit* %arg1, %Qubit* %arg2)
}
; Assuming main is the entry-point function.
define void @main() {
...
    call void @swap(%Qubit* null, %Qubit* nonnull inttoptr (i64 1 to %Qubit*))
...
}
```

**Invalid example**:
```llvm
define void @swap(%Qubit* %arg1, %Qubit* %arg2) {
    call void __quantum__qis__cnot__body(%Qubit* %arg1, %Qubit* %arg2)
    call void __quantum__qis__cnot__body(%Qubit* %arg2, %Qubit* %arg1)
    call void __quantum__qis__cnot__body(%Qubit* %arg1, %Qubit* %arg2)
}
; Assuming main is the entry-point function.
define void @main() {
...
    ; This is invalid because the `%Qubit*` arguments used when calling the user-defined function are not constants.
    %0 = __quantum__rt__qubit_allocate()
    %1 = __quantum__rt__qubit_allocate()
    call void @swap(%Qubit* %1, %Qubit* %2)
...
}
```

3. If called from within another user-defined function, the `%Qubit*%` arguments it receives can only be arguments from the caller function.

 **Valid example**:
```llvm
define void @swap(%Qubit* %arg1, %Qubit* %arg2) {
    call void __quantum__qis__cnot__body(%Qubit* %arg1, %Qubit* %arg2)
    call void __quantum__qis__cnot__body(%Qubit* %arg2, %Qubit* %arg1)
    call void __quantum__qis__cnot__body(%Qubit* %arg1, %Qubit* %arg2)
}

define void @swap_and_h(%Qubit* %arg1, %Qubit* %arg2) {
    call void @swap(%Qubit* %arg1, %Qubit* %arg2)
    call void __quantum__qis__h(%Qubit* %arg1)
    call void __quantum__qis__h(%Qubit* %arg2)
}
; Assuming main is the entry-point function.
define void @main() {
...
    call void @swap(%Qubit* null, %Qubit* nonnull inttoptr (i64 1 to %Qubit*))
    call void @swap(%Qubit* null, %Qubit* nonnull inttoptr (i64 1 to %Qubit*))
...
}
```

**Invalid example**:
```llvm
define void @swap(%Qubit* %arg1, %Qubit* %arg2) {
    call void __quantum__qis__cnot__body(%Qubit* %arg1, %Qubit* %arg2)
    call void __quantum__qis__cnot__body(%Qubit* %arg2, %Qubit* %arg1)
    call void __quantum__qis__cnot__body(%Qubit* %arg1, %Qubit* %arg2)
}

define void @swap_and_h(%Qubit* %arg1, %Qubit* %arg2) {
    ; Invalid because it calls a user-defined function from another user-defined function using constants and dynamically allocated qubits as arguments.
    %1 = __quantum__rt__qubit_allocate()
    call void @swap(%Qubit*  null, %Qubit* %1)
    call void __quantum__qis__h(%Qubit* %arg1)
    call void __quantum__qis__h(%Qubit* %arg2)
}
; Assuming main is the entry-point function.
define void @main() {
...
    call void @swap(%Qubit* null, %Qubit* nonnull inttoptr (i64 1 to %Qubit*))
    call void @swap(%Qubit* null, %Qubit* nonnull inttoptr (i64 1 to %Qubit*))
...
}
```